CRM/Mailing/BAO/Mailing.php

```diff
@@ -367,6 +377,8 @@ public static function getRecipients($mailingID) {
     $mailingGroup->reset();
     $mailingGroup->query(" DROP TEMPORARY TABLE $excludeTempTablename ");
     $mailingGroup->query(" DROP TEMPORARY TABLE $includedTempTablename ");
+
+    CRM_Utils_Hook::alterMailingRecipients($mailingObj, $params, 'post');
```

Comment:
Continuing thoughts on contract, the second question: *What are you supposed to do in `post`?*

Is the idea that one would directly `INSERT`/`DELETE` records from `civicrm_mailing_recipients`?

Reply:
As a consumer, I guess I can see how that's kind'a nice -- i.e. *you don't have think at all about how the main query-building works; you can just do your own thing*.

OTOH, recall this: the CiviMail UI shows a preview of the #recipients, and it actually executes `getRecipients()` but performs a `ROLLBACK`. This was sort of a quick hack that people have periodically complained about, but it was easier than rewriting `getRecipients()`. Can we ever get rid of that hack?

As of today, the answer is, "Yes, if we spend a while refactoring."

If we change the contract so that `hook_civicrm_alterMailingRecipients` relies on direct manipulation of the `civicrm_mailing_recipients` table, then the answer is "No, we can't cleanup the weird ROLLBACK without breaking the contract."